File: native/app/contribution_read.go
package app

import (
	"errors"
	"github.com/JohnnyPBelo/relayloom/native/core"
	"github.com/JohnnyPBelo/relayloom/native/sites"
	"time"
)

func (n *Node) contributionFormLocked(body map[string]any) (any, error) {
	lookup, err := sites.ParseContributionFormLookup(body)
	if err != nil {
		return nil, err
	}
	if n.identity == nil {
		return nil, errors.New("desbloqueie a identidade")
	}
	if _, err = n.objectsLocked(); err != nil {
		return nil, err
	}
	bundle, err := n.Store.GetWithTouch(text(lookup["snapshotId"]), false)
	if err != nil {
		return nil, err
	}
	if contains(n.config.Blocked, bundle.Manifest.Author.ID) {
		return nil, errors.New("autor do site bloqueado")
	}
	if m, ok := n.private.Mutations[bundle.Manifest.ID]; ok && m.Author == bundle.Manifest.Author.ID && m.Deleted && m.Expires > time.Now().UnixMilli() {
		return nil, errors.New("o autor retirou este snapshot")
	}
	value, err := core.DecryptBundle(bundle, n.identity)
	if err != nil {
		return nil, err
	}
	found, err := sites.ResolveContributionForm(lookup, bundle, value, n.identity.Public.ID, time.Now().UnixMilli())
	if err != nil {
		return nil, err
	}
	return sites.DescribeContributionForm(found)
}

File: apps/node/src/contribution-read.ts
import {
  decryptBundle,
  type Identity,
  type ContentStore,
} from "../../../packages/core/src/index";
import { nodeCertificateCrypto } from "../../../packages/core/src/certificate-crypto";
import {
  parseContributionFormLookup,
  createContributionContextResolver,
} from "../../../packages/sites/src/contribution-context";
const resolver = createContributionContextResolver(nodeCertificateCrypto);
export function readContributionForm(
  input: unknown,
  context: {
    identity: Identity;
    store: ContentStore;
    blocked(): readonly string[];
    withdrawn(id: string, authorId: string): boolean;
  },
) {
  const q = parseContributionFormLookup(input),
    bundle = context.store.get(q.snapshotId, false),
    owner = bundle.manifest.author.id;
  if (context.blocked().includes(owner)) throw Error("Autor do site bloqueado");
  if (context.withdrawn(bundle.manifest.id, owner))
    throw Error("O autor retirou este snapshot");
  const plaintext = decryptBundle(bundle, context.identity);
  const found = resolver.resolve(
    q,
    bundle,
    plaintext,
    context.identity.public.id,
    Date.now(),
  );
  return resolver.describe(found);
}
